pulse2percept/retina.py

```diff
@@ -1,772 +1,958 @@
+            Dimensions: <#electrodes x #time points>
+        layers : list
+            List of retinal layers to simulate. Choose from:
+            - 'OFL': optic fiber layer
+            - 'GCL': ganglion cell layer
+            - 'INL': inner nuclear layer
+        """
+        not_supported = np.array([l not in SUPPORTED_LAYERS for l in layers],
+                                 dtype=bool)
+        if any(not_supported):
+            raise ValueError("Acceptable values for `layers` is 'OFL', 'GCL', "
+                             "'INL'.")
+
+        ecm = np.zeros((ecs_item.shape[0], pt_list[0].shape[-1]))
+        if 'INL' in layers:
+            ecm[0, :] = np.sum(ecs_item[0, :, np.newaxis] * pt_list, axis=0)
+        if ('GCL' or 'OFL') in layers:
+            ecm[1, :] = np.sum(ecs_item[1, :, np.newaxis] * pt_list, axis=0)
+        return ecm
+
+    def model_cascade(self, ecs_item, pt_list, layers, dojit):
+        """The Temporal Sensitivity model
+
+        This function applies the model of temporal sensitivity to a single
+        retinal cell (i.e., a pixel). The model is inspired by Nanduri
+        et al. (2012), with some extended functionality.
+
+        Parameters
+        ----------
+        ecs_item: array-like
+            A 2D array specifying the effective current values at a particular
+            spatial location (pixel); one value per retinal layer and
+            electrode.
+            Dimensions: <#layers x #electrodes>
+        pt_list: list
+            A list of PulseTrain `data` containers.
+            Dimensions: <#electrodes x #time points>
+        layers : list
+            List of retinal layers to simulate. Choose from:
+            - 'OFL': optic fiber layer
+            - 'GCL': ganglion cell layer
+            - 'INL': inner nuclear layer
+        dojit : bool
+            If True, applies just-in-time (JIT) compilation to expensive
+            computations for additional speed-up (requires Numba).
+
+        Returns
+        -------
+        Brightness response over time. In Nanduri et al. (2012), the
+        maximum value of this signal was used to represent the perceptual
+        brightness of a particular location in space, B(r).
+        """
+        # For each layer in the model, scale the pulse train data with the
+        # effective current:
+        ecm = self.calc_layer_current(ecs_item, pt_list, layers)
+
+        # Calculate charge accumulation on the input
+        ca = self.charge_accumulation(ecm)
+
+        # Sparse convolution is faster if input is sparse. This is true for
+        # the first convolution in the cascade, but not for subsequent ones.
+        if 'INL' in layers:
+            fr_inl = self.fast_response(ecm[0], self.gamma_inl,
+                                        dojit=dojit,
+                                        method='sparse')
+
+            # Cathodic and anodic parts are treated separately: They have the
```

Comment:
Since this repeats exactly below, maybe we can make a little helper function out of it?